### DISCUSSION TITLE
Our queries and mutation result slows down with time

### DISCUSSION THREAD

**Discussion by ahmed-trailfive** [prisma/prisma#14491] [Q&A]
Hi, need some help while deploying the Prisma project. 

**Our queries and mutation result slows down with time**

We are using the following technologies in our API

- Node
- Prisma
- Apollo Server
- MySQL (docker)
Our Queries and mutations are simple.

**Where should I start debugging from?**

Our server is deployed on the VPS with 60GB RAM.

what I have tried so far 
https://www.prisma.io/docs/guides/performance-and-optimization 
MySQL parameter optimization.

**Comment by ahmed-trailfive** [comment]
The queries slow down even on localhost.

**Reply by ahmed-trailfive** [reply]
I think this time is increasing as I am using big "include" statements. with every Query.

**Comment by austincrim** [comment]
Hey there 👋,

Are you saying that your queries get progressively slower during the lifetime of your application or are they always slow?

How are you measuring the queries?

**Reply by ahmed-trailfive** [reply]
during the life of application as data increases. Data is not too much but in-depth relations exist. 
for example, Project includes a track, tracks include multiple recordings, recordings include files, etc.
A simple query like this 
![image](https://user-images.githubusercontent.com/97940796/181168151-29ded470-47ac-459d-8b51-f231fda5269e.png)

Have a look at the performance matrics. 
![image](https://user-images.githubusercontent.com/97940796/181169039-373cf2ed-b1ba-4250-8f47-45a9de809f03.png)

**Reply by ahmed-trailfive** [reply]
The duration is the output of the following code
```export const prisma = new PrismaClient({
    log: [
        {
            emit: 'event',
            level: 'query',
        },
        {
            emit: 'stdout',
            level: 'error',
        },
        {
            emit: 'stdout',
            level: 'info',
        },
        {
            emit: 'stdout',
            level: 'warn',
        },
    ],
})
prisma.$on('query', (e: any) => {
    console.log('Duration: ' + e.duration + 'ms')
})
```
But as you can see only 10 DB operations with durations in milliseconds and query total operations are 7681

**Reply by ahmed-trailfive** [reply]
Not able to understand why the  ```query_total_operations``` and ```query_total_queries``` are too high. These parameters were reaching 80k to 90k and choking the system when I was including all the fields in the ```include```

**Reply by ahmed-trailfive** [reply]
We are using the apollo Studio too.

**Reply by ahmed-trailfive** [reply]
after removing much include statements the dB operations are reduced to less than 500 
![image](https://user-images.githubusercontent.com/97940796/181180634-595fedfd-a80f-49d1-8a54-1bb5d9bf6a9a.png)

**Reply by ahmed-trailfive** [reply]
This thing means that the include does not solves the n+1 problem or I have too much in-depth relation https://www.prisma.io/docs/guides/performance-and-optimization/query-optimization-performance#solving-n1-with-include

**Reply by ahmed-trailfive** [reply]
I know Prisma include does solve the n+ 1 problem

**Reply by jgb-solutions** [reply]
@ahmed-trailfive sorry to ask but how fast the queries used to be before that? I mean less than 10ms is pretty fast to me.

**Reply by ahmed-trailfive** [reply]
Hi! @jgb-solutions Thank you for asking. previously I was including all the relations in the include as I thought there is no need to work with ```include``` as it handles n+1. The queries with the passage of time start to take 1 to 3 sec to respond as the data increases. As we diagnosed that too many SQL queries are being generated. So we reduced the fields in the ```include``` and this helped.

But One issue is still stuck in my mind.
Now the front-end devs have to ask me if they need the following field :like 
they say: we need the Project that returns the {projects(model)->tracks(model)->recordings(model)>files(model)>urls(string)}
So now I have to manually include the above-mentioned requirement. and I think this violates the GraphQL philosophy that "front-end can request whatever they want."

**Reply by jgb-solutions** [reply]
@ahmed-trailfive oh I see. Yeah 1-3s is very slow for database queries. I saw GraphQL support for N+1 queries in the docs. Does any of this help? https://www.prisma.io/docs/guides/performance-and-optimization/query-optimization-performance#solving-n1-in-graphql-with-findunique-and-prismas-dataloader

**Reply by austincrim** [reply]
Could you share your full schema and how much data you are querying?

**Reply by ahmed-trailfive** [reply]
Sorry @austincrim due to privacy issues I cannot share the schema file however I can specify the following information

```
some self relations as well like derived from which Projects
model Project{
 43 attributes
 20 relations
 }
 
 model Track{
  12 attributes
  9 relations
}

model Recording{
  27 attributes
  10 relations
}
model File{
  10 attributes
  2 relations
}
```
The application mostly fetches the data in the following way
```
Project{
        2 to 5 fields
       Track{ //there can be multiple tracks
             2 to 5 fields
             Recording{  // there can be multiple recordings
                  2 to 5 fields
                  File{   // there can be multiple files.
                       2 to 5 fields
                        url
                  }
            }
      }
 }

**Reply by ahmed-trailfive** [reply]
@austincrim Actually what comes to my mind is that the ORM should only get the requested data by front-end from DB. For example, we get the requested data and pass it to include a statement. Just an opinion. What's your say on this?

**Reply by ahmed-trailfive** [reply]
found a similar question https://stackoverflow.com/questions/71650767/how-can-i-use-the-fields-in-a-graphql-query-to-perform-nested-reads-with-prisma

**Reply by ahmed-trailfive** [reply]
@austincrim

**Reply by austincrim** [reply]
Could you provide a full reproduction repository with your setup? It is difficult to reproduce or understand your issue without seeing the whole picture of your app.

**Reply by vladi-stevanovic** [reply]
👋  Hello @ahmed-trailfive I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! 🙌

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
